drivers/flash/flash_stm32_ospi.c

```diff
@@ -519,14 +519,51 @@ static bool ospi_address_is_valid(const struct device *dev, off_t addr,
 	return (addr >= 0) && ((uint64_t)addr + (uint64_t)size <= flash_size);
 }
 
+static int stm32_ospi_wait_auto_polling(struct flash_stm32_ospi_data *dev_data,
+		OSPI_AutoPollingTypeDef *s_config, uint32_t timeout_ms)
+{
+	/* If interrupt is enabled, no need for busy waiting to check the poll result
+	 * like HAL_OSPI_AutoPolling() does.
+	 */
+	if (irq_is_enabled(DT_IRQN(STM32_OSPI_NODE))) {
+		dev_data->cmd_status = 0;
+		if (HAL_OSPI_AutoPolling_IT(&dev_data->hospi, s_config) != HAL_OK) {
+			LOG_ERR("OSPI AutoPoll failed");
+			return -EIO;
+		}
+		if (k_sem_take(&dev_data->sync, K_MSEC(timeout_ms)) != 0) {
+			LOG_ERR("OSPI AutoPoll wait failed");
+			HAL_OSPI_Abort(&dev_data->hospi);
+			/* Ensure the semaphore is left value 0 even if the interrupt callback
+			 * is executed after the expired wait above.
+			 */
+			k_sem_reset(&dev_data->sync);
+			return -EIO;
+		}
+
+		/* HAL_OSPI_AutoPolling_IT enables transfer error interrupt which sets
+		 * cmd_status.
+		 */
+		return dev_data->cmd_status;
+
+	} else if (HAL_OSPI_AutoPolling(&dev_data->hospi, s_config, timeout_ms) != HAL_OK) {
+		LOG_ERR("OSPI AutoPoll failed");
+		return -EIO;
+	}
+
+	return 0;
+}
+
 /*
  * This function Polls the WEL (write enable latch) bit to become to 0
  * When the Chip Erase Cycle is completed, the Write Enable Latch (WEL) bit is cleared.
  * in nor_mode SPI/OPI OSPI_SPI_MODE or OSPI_OPI_MODE
  * and nor_rate transfer STR/DTR OSPI_STR_TRANSFER or OSPI_DTR_TRANSFER
  */
-static int stm32_ospi_mem_erased(OSPI_HandleTypeDef *hospi, uint8_t nor_mode, uint8_t nor_rate)
+static int stm32_ospi_mem_erased(struct flash_stm32_ospi_data *dev_data,
```

Comment:
Yes I think it could. If you see passing quite generic `dev` pointer to these functions is not too generic, I could take a look to refactor these.

Reply:
`stm32_ospi_mem_ready` is used also on other modes than in the one set in `dev_cfg->data_mode` during the initialization before the mode is changed to the target one. At least I see it's this way. The same applies to `stm32_ospi_write_enable`: parameters can be something else than the final mode/rate. 

`stm32_ospi_mem_erased` could be easily changed to use `dev` pointer since erase operations are not done before changing to the target mode.